Re-introduce the OPTIONAL_MEMORIAL rank

emagnier opened the pull request:
**Reintroduction of the `OPTIONAL_MEMORIAL` rank to:**
- Distinguish obligatory memorials from optional memorials using the `rank` property.
- It is still possible to distinguish them using the `precedence` or `isOptional` properties.
- Provide specific localization for the `OPTIONAL_MEMORIAL` value.

**Notes:**
- Up until Romcal v2, this value was titled `OPT_MEMORIAL`. I prefer to state it without abbreviation for clarity.
- The translations here come from the git history of this project.
- All the `rank` translations are now all in lowercase.

Resolves https://github.com/romcal/romcal/issues/435.

emagnier committed:
feat!: re-introduce the OPTIONAL_MEMORIAL rank

tukusejssirs commented:
> All the rank translations are now all in lowercase.

While I agree with you, @emagnier, I’d like to emphasise that this should follow grammatical rules of a given locale, e.g. in German (not yet supported by romcal), all nouns are written with capital initial letter, therefore we should not assume the ranks has to be written in lowercase. That said, we should prefer lowercase to uppercase.

tukusejssirs commented on lib/constants/ranks.ts:
```diff
@@ -31,16 +31,29 @@ export enum Ranks {
-   * Optional Memorials.
-   * If several Optional Memorials are inscribed in the Calendar on the same day, only
-   * one may be celebrated, the others being omitted. (UNLY #14)
+   * Obligatory Memorials are liturgical commemorations of saints, events, or aspects of the faith.
```
In general, `obligatory memorial` (and any other celebration rank) is not a proper noun in English, therefore, it should not be written with capital initial letters, unless it is part of a title (heading, celebration name, etc).

We might want to move this to a separate PR, as this should be fixed in all files and I have no problem with doing so. Therefore, consider this comment as non-blocking. :wink:

```suggestion
   * Obligatory memorials are liturgical commemorations of saints, events, or aspects of the faith.
```

emagnier replied:
I like the convention of capitalizing these names to emphasize the importance of these names, which are the subject of the description here. This writing rule should only be used in such cases (for describing a property or an acronym, for example).

However, at the same time, and for the sake of consistency, I understand your point and your preference for following the general writing rules.

Before choosing one writing style over the other, what are your thoughts on this?

tukusejssirs replied:
> to emphasize the importance of these names

Based on what you what to accomplish, I’d choose one of the following:
- reference a type: use [TSDoc links](https://tsdoc.org/pages/tags/link/). e.g. `{@link Ranks.OptionalMemorial}` (once per TSDoc block);
- emphasise some strings: use Markdown syntax for bold and/or italics;
- emphasise some code: use Markdown syntax for inline code snippets.

I’d always use Markdown syntax in TSDocs anyway. I’d also try to use grammatically correct code comments in simple English.

> acronym

I think we need to create a glossary or a list of terms where we would define some of the most commonly used stuff, including acronyms and shortenings (which would link to an entry with full name via `see $link`). Then, we could use `{@links $link}` in TSDocs when needed.

emagnier replied:
Okay, and thank you for your recommendations. I will update these descriptions following your suggestions 👍

emagnier replied:
> I think we need to create a glossary or a list of terms where we would define some of the most commonly used stuff

Indeed, this is clearly something I would like to see added to Romcal.

tukusejssirs commented on lib/constants/ranks.ts:
```diff
@@ -31,16 +31,29 @@ export enum Ranks {
+   * following the liturgical norms outlined in the General Instruction of the Roman Missal and the
+   * Liturgy of the Hours.
+   * When an Obligatory Memorial falls on a weekday during the liturgical season of Lent or a
+   * privileged weekday of Advent, it must only be celebrated as an Optional Memorial, as Lent and
```
Aren’t memorials in Lent and Advent (after 17 Dec including) celebrated rather as _commemorations_?

And again, `obligatory memorial` is not a proper noun in English _per se_.

emagnier replied:
Would you think that in these cases (during Lent and from 17 Dec), we should define a new `rank` type as `COMMEMORATION`? At first glance, I think it's a good idea  👍  Especially since the precedence property also provides useful information about the celebration.

tukusejssirs replied:
Well, defining a new `rank` for commemorations is a good idea, however, I am not sure if that is that necessary (there is no such type in the GIRM). IMHO adding a `LiturgicalDay` (celebration) property (e.g. `isCommemoration` of a `boolean` type) could suffice.

emagnier replied:
I'm not very enthusiastic about adding an `isCommemoration` property. 
Especially since memories, feasts, and other ranks are also commemorations, having an `isCommemoration: false` property for these celebrations seems counterintuitive to me. 

The term "commemoration" (for a rank) remains quite generic, so in our case here, it seemed appropriate (i.e. we specify that this element remains a commemoration, even though it no longer holds the official rank of a memory).

Let's save this discussion for later and for a dedicated topic because I believe we need to reexamine the missal instructions on this issue and let this idea mature a bit.

tukusejssirs replied:
> I'm not very enthusiastic about adding an `isCommemoration` property.
> Especially since memories, feasts, and other ranks are also commemorations, having an `isCommemoration: false` property for these celebrations seems counterintuitive to me.

You might be right.

I was referring to this section of the GIRM:

> 

On the weekdays of Advent from December 17 to December 24, on days within the Octave of Christmas, and on the weekdays of Lent, except Ash Wednesday and during Holy Week, the Mass for the current liturgical day is to be used; but the collect may be taken from a memorial which happens to be listed in the General Calendar for that day, except on Ash Wednesday and during Holy Week. On weekdays of the Easter Season, memorials of Saints may rightly be celebrated fully.

Whenever a proper celebration is prohibited by this and only the _collect may be take from a memorial_, I considered this a _commemoration_. I have no idea where I found this term though.

> Let's save this discussion for later and for a dedicated topic because I believe we need to reexamine the missal instructions on this issue and let this idea mature a bit.

Agreed. :+1:

tukusejssirs commented on lib/constants/ranks.ts:
```diff
@@ -31,16 +31,29 @@ export enum Ranks {
+   * Their observance is mandatory and integrated into the celebration of the occurring weekday,
+   * following the liturgical norms outlined in the General Instruction of the Roman Missal and the
+   * Liturgy of the Hours.
+   * When an Obligatory Memorial falls on a weekday during the liturgical season of Lent or a
```
```suggestion
   * When an obligatory memorial falls on a weekday during the liturgical season of Lent or a
```

tukusejssirs commented on lib/constants/ranks.ts:
```diff
@@ -31,16 +31,29 @@ export enum Ranks {
   Memorial = 'MEMORIAL',
 
+  /**
+   * Optional Memorials are liturgical commemorations of saints, events, or aspects of the faith,
```
```suggestion
   * Optional memorials are liturgical commemorations of saints, events, or aspects of the faith,
```

tukusejssirs commented on lib/constants/ranks.ts:
```diff
@@ -31,16 +31,29 @@ export enum Ranks {
+   * Their observance is integrated into the celebration of the occurring weekday, adhering to the
+   * liturgical norms provided in the General Instruction of the Roman Missal and the Liturgy of
+   * the Hours.
+   * In cases where multiple Optional Memorials are designated on the same day in the liturgical
```
```suggestion
   * In cases where multiple optional memorials are designated on the same day in the liturgical
```

tukusejssirs commented on lib/constants/ranks.ts:
```diff
@@ -31,16 +31,29 @@ export enum Ranks {
+   * the Hours.
+   * In cases where multiple Optional Memorials are designated on the same day in the liturgical
+   * calendar, only one of them may be celebrated, and the others must be omitted (UNLY #14).
+   * This allows for some flexibility in choosing which Optional Memorial to commemorate when
```
```suggestion
   * This allows for some flexibility in choosing which optional memorial to commemorate when
```

emagnier commented:
If the rule in a language is to keep these names capitalized, I am, of course, 100% in favor of following that rule. Otherwise, indeed, I also think it's better to keep them in lowercase as it makes them easier to use in string composition.

tukusejssirs commented on lib/types/locale.ts:
```diff
@@ -52,6 +52,7 @@ export interface Locale {
     sunday?: string;
     feast?: string;
     memorial?: string;
+    optional_memorial?: string;
     weekday?: string;
   };
```
How about making `ranks`, `colors`, `weekdays` and `months` required? IMHO, they can be easily translated in any locale, as these strings are very common.

emagnier replied:
No, because an XX-XX locale may extend an XX locale. So, at this level, it should remain optional (the script added here #461 will, however, ensure that all these items are correctly defined and present for a language).

tukusejssirs replied:
You’re right, I have forgotten about the regional locales. :+1:

tukusejssirs reviewed (approved):
@emagnier, thanks for the PR. It LGTM, I have just some (regular) nitpicks. :wink:

If you want, you merge the PR as is.

emagnier committed:
docs: update rank descriptions

emagnier commented:
@tukusejssirs, FYI I have rebased the PR, and added a commit to document the breaking change in the changelog (while waiting for the changelog update to be automated someday).

emagnier committed:
docs: specify breaking change to the changelog

tukusejssirs commented on CHANGELOG.md:
```diff
@@ -8,7 +8,15 @@ The romcal v3 is a major rewrite of the core library, with performance and archi
+
+#### Breaking change
+
+- The `OPTIONAL_MEMORIAL` `rank` is being reintroduced, which was previously available as `OPT_MEMORIAL` until Romcal v2. This re-introduction includes related localizations for all currently supported locales.
```
I probably use `rank` here as a term not as a code. In case you still want to use it as a code, IMHO it should be written as `Rank`.

```suggestion
- The `OPTIONAL_MEMORIAL` rank is being reintroduced, which was previously available as `OPT_MEMORIAL` until Romcal v2. This re-introduction includes related localizations for all currently supported locales.
```

emagnier replied:
`rank` was meant to refer to the `rank` property, not the `Rank` type. But alright, I'm going with your suggestion 👍👍. Anyway, you have a great eye for detail (which is excellent), but let's merge this PR now and continue to make progress on other matters.

tukusejssirs reviewed (approved):
I have a minor nitpick, however, it is nothing serious, therefore I approve the PR in case you want to merge it anyway. :wink: :pray:

emagnier committed:
Update CHANGELOG.md

Co-authored-by: Tukusej’s Sirs <tukusejssirs@protonmail.com>
